README.MD
# cofragr: cfDNA co-fragmentation analysis in R

<details open="open">
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#understanding-cofragr">Understanding cofragr</a></li>
    <li>
      <a href="#usage">Usage</a>
      <ul>
        <li><a href="#preparation">Preparation</a></li>
        <li><a href="#analysis-with-rscript">Analysis with RScript</a></li>
        <li><a href="#analysis-with-snakemake">Analysis with Snakemake</a></li>
      </ul>
    </li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
  </ol>
</details>

## About The Project

cofragr is an R package for cofragmentation patterns of cell-free DNA.

## Getting Started

### Prerequisites

The cofragr package is still under development and hasn’t been released to CRAN, so you must install cofragr by using the **devtools** workflow. You can install **devtools** two different ways:

You can install the library directly from CRAN:

```R
install.packages("devtools")
```

You can also install the development version from GitHub:

```R
devtools::install_github("r-lib/devtools")
```

You can learn more about **devtools** and get help with its installation [here](https://devtools.r-lib.org/).


### Installation

Now that you've downloaded **devtools**, you can use the following command to install the package:

```R
devtools::install_github("epifluidlab/cofragr")
```

Congratulations! You've successfully downloaded and installed **cofragr**!

*Note: If you're having any problems installing cofragr, make sure that you have the latest version of R installed.*

## Understanding cofragr

Understanding how **cofragr** works can be crucial to its successful usage. There are three core functions to this R package:

##### read_fragments

Calling this function will read the fragment data file, which is typically a BED file. Since this functions uses the package **[bedtools](https://haizi-zh.github.io/bedtorch/)** to read the fragment files, the arguments/options for `range` and `genome` will follow the same format. 

###### Options

```
cofragr::read_fragments(file_path, range = NULL, genome = NULL)
```

`range; default: range = NULL`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the ranges of the BED file that read_fragments will read.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Note that `range` follows the standard genomic range notation, ex. "chr1:1001-2000". 

`genome; default: genome = NULL`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the reference genome for the BED file in question.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Note that `genome` can take any of the `genome` arguments that [`GenomeInfoDb::Seqinfo()`](https://rdrr.io/bioc/GenomeInfoDb/man/Seqinfo-class.html) can.

Again, you can read more about the options behind `range` and `genome` [here](https://haizi-zh.github.io/bedtorch/reference/read_bed.html).

##### contact_matrix

Calling this function will calculate the contact matrix from the fragment data file, which should have already been called by the `read_fragments` function. 

###### Options

```
cofragr::contact_matrix(frag, bin_size = 500e3L, n_workers = 1L, subsample = 10e3L, min_sample_size = 100L, bootstrap = 1L, seed = NULL)
```
Note that `frag` is already defined as the fragment file as a result of the `read_fragments` function.

`bin_size; default: bin_size = 500e3L`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Specifies the size of the bins for the contact matrix in base-pair units.

`n_workers; default: n_workers = 1L`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Specifies the number of workers for the task at hand.

`subsample; default: subsample = 10e3L`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Specifies the sampling size for the contact matrix calculations. 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Note:* In this case, the subsample means for each genomic bin, we only randomly select 1 in every 10,000 fragments for the analysis. This counteracts the fact that the number of fragments are different between different bins.

`bootstrap; default: bootstrap = 1L`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Specifies the number of bootstrap samples in the contact matrix calculations.

##### write_contact_matrix

Calling this function will write the contact matrix to a specified filepath; the contact matrix should have already been calculated by the `contact_matrix` function. 

###### Options

```
cofragr::write_contact_matrix(cm, file_path, comments = NULL)
```

`comments; default: comments = NULL`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Allows the addition of a character vector to be appended at the top of the written BED file as a header.

## Usage

cofragr provides both R APIs and an R script for the analysis. 

### Preparation

The pipeline requires fragment data files as input. A fragment data file is essentially a BED file, with each row representing a cfDNA fragment. Below is an example:

    14      19000035        19000198        .       27      +
    14      19000044        19000202        .       42      +
    14      19000045        19000202        .       20      -
    14      19000049        19000202        .       12      +

The six columns are: chrom, start, end, feature name (which we do not use), MAPQ (which is the smallest MAPQ of the two paired reads), and strand (which we do not use).

You can easily prepare the fragment data file from a query-sorted BAM file, or refer to our FinaleDB paper for more details (Zheng, Zhu, and Liu 2020).

### Analysis with RScript

The quickest way to start is using the **RScript** for your dataset.

You can find the script in R package installation:

    system.file("extdata/scripts", "cofragr.R", package = "cofragr")       

Or simply clone the source code from GitHub and find the script in the directory `inst/ext/scripts/`.


Then you can simply run some variation of the following, inputting any arguments as you desire.

    Rscript cofragr.R 
    -i examplefile.bed.gz \
    -o output_file_path \
    --min-mapq 30 \
    --chroms 1:2:3 \
    --subsample 10000 \
    --min-fraglen 50 \
    --max-fraglen 350 \
    --sample-id example \

We encourage you to look through the script file, `cofragr.R`, to understand all of the available arguments.

##### Output

You will be outputted two files, the calculated contact matrix and its accompanying index file. 

cofrag_cm.bed.gz is the calculated contact matrix:

    #chrom  start   end     chrom2  start2  end2    score   n_frag1 n_frag2 p_value p_value_sd
    1       0       500000  1       0       500000  15.14129840182        435     435     0.58172039812       .
    1       0       500000  1       500000  1000000 0       435     2539    0       .
    1       0       500000  1       1000000 1500000 0       435     3997    0       .
    1       0       500000  1       1500000 2000000 0       435     3707    0       .
    1       0       500000  1       2000000 2500000 0       435     4361    0       .


### Analysis with Snakemake

For large-scale real applications, we suggest use workflow management tools such as **snakemake**, nextflow, etc. We provide a snakemake file. Similar to cofragr.R, you can find it:

In R package installation:

    system.file("extdata/scripts", "cofragr.smk", package = "cofragr")

Or simply clone the source code from GitHub and find the script in the directory `inst/ext/scripts/`.

*Note: Usually snakemake workflow is more related to your specific computing environment. As a result, the example above is only for your reference. Have a look in both the snakemake file and the command above, and modify it accordingly.*

## License

See `LICENSE` for more information.

## Contact

* Haizi Zheng haizi.zh@gmail.com
* Yaping Liu lyping1986@gmail.com
* Ravi Bandaru ravi14.bandaru@gmail.com


CHANGELOG.md
## 2021-10-17

Version: v0.1.2.1

New:

* Support GRCh37 and GRCh38

Changes:

* Optimize codes


tests/testthat.R
library(testthat)
library(cofragr)

test_check("cofragr")


R/cofragr-package.R
# R package for cofragmentation analysis

#' @importFrom dplyr `%>%` tibble as_tibble mutate rename select filter group_by ungroup summarize arrange inner_join lag
#' @importFrom stringr str_interp str_split str_detect str_sort str_match
#' @importFrom purrr map2_dfr map_dbl map
#' @importFrom tidyr expand_grid
#' @importFrom magrittr `%<>%`
#' @importFrom GenomicRanges makeGRangesFromDataFrame findOverlaps mcols `mcols<-` pintersect ranges `ranges<-` width `width<-` seqnames start `start<-` end `end<-` seqinfo `seqinfo<-`
#' @importFrom S4Vectors Rle DataFrame from to
#' @importFrom parallel makeCluster makeForkCluster stopCluster
#' @importFrom doParallel registerDoParallel
#' @importFrom doRNG `%dorng%`
#' @importFrom foreach foreach `%do%` `%dopar%`
#' @importFrom rlang is_null
#' @importFrom assertthat assert_that
NULL


inst/extdata/scripts/cofragr_comp.R
#!/usr/bin/env Rscript

stop_quietly <- function() {
  opt <- options(show.error.messages = FALSE)
  on.exit(options(opt))
  stop()
}


# # example
# script_args <- list(
#   input = "~/Downloads/EE87929.cofrag_cm.bed.gz",
#   output_dir = here::here("sandbox/"),
#   sample_id = "EE87929",
#   res = 500e3L,
#   chroms = c("20", "21", "22"),
#   # method = c("juicer", "lieberman", "obs_exp"),
#   method = c("lieberman", "obs_exp"),
#   smooth = c(1L, 3L),
#   standard_compartment = "wbc",
#   genome = "hs37-1kg",
#   juicer = NULL,
#   java = "java"
# )


# script_args validity
validate_args <- function(args) {
  with(args, {
    # assert_that(
    #   is_null(standard_compartment) ||
    #     (
    #       is_scalar_character(standard_compartment) &&
    #         standard_compartment %in% c("wbc", "gene_density")
    #     )
    # )
    assert_that(is_scalar_integer(res) && res > 0)
    assert_that(is_scalar_character(genome) && genome %in% c("GRCh38", "hs37-1kg"))
    assert_that(length(method) >= 1 &&
                  all(method %in% c("juicer", "lieberman", "obs_exp")))
    # assert_that(length(smooth) >= 1 && is_integer(smooth) && all(smooth > 0))
  }) %>% invisible()
} 


if (interactive()) {
  library(tidyverse)
  library(magrittr)
  library(here)
  library(rlang)
  library(assertthat)

  if (is.null(get0("script_args")))
    stop_quietly()
} else {
  library(here)
  library(rlang)
  library(assertthat)
  
  # Run in CLI script mode
  parser <- optparse::OptionParser(
    option_list = list(
      optparse::make_option(c("-i", "--input")),
      optparse::make_option(c("-o", "--output-dir")),
      optparse::make_option(c("-s", "--sample-id")),
      optparse::make_option(c("--res"), type = "integer", default = 500e3L),
      optparse::make_option(c("--chroms"), default = NULL,
                            help = "Perform the analysis only for a selected group of chromosomes. Separated by colons, such as 12:16:X. If not provided, all chromosomes found in the input file will be used"),
      optparse::make_option(c("--method"), default = "juicer:lieberman:obs_exp"),
      optparse::make_option(c("--smooth"), default = "1:3", help = "Apply moving average on the compartment track. Sometimes this can remove some quirks and make the data more similar to Hi-C compartment tracks"),
      optparse::make_option(
        c("--standard-compartment"),
        default = NULL,
        help = "A standard compartment track (BED format) for calculating compartment correlation scores. Can be either `wbc` or `gene_density`"
      ),
      optparse::make_option(
        c("-g", "--genome"),
        default = NULL,
        help = "Reference genome of the dataset. The default is hs37-1kg.",
      ),
      optparse::make_option(c("--juicer"), default = NULL,
                            help = "Path to the .jar file Juicer tools. If not provided, will download from Internet"),
      optparse::make_option(c("--java"), default = "java",
                            help = "Path to java")
    )
  )
  script_args <-
    optparse::parse_args(
      parser,
      args = commandArgs(trailingOnly = TRUE),
      convert_hyphens_to_underscores = TRUE
    )

  library(tidyverse)
  library(magrittr)

  script_args$method %<>% str_split(pattern = ":") %>% .[[1]]
  script_args$smooth %<>% str_split(pattern = ":") %>% .[[1]] %>% map_int(as.integer)
  if (!is_null(script_args$chroms))
    script_args$chroms %<>% str_split(pattern = ":") %>% .[[1]]
}

validate_args(script_args)

# Build comment lines
comments <- c(
  paste0("cofragr_version=", as.character(packageVersion("cofragr"))),
  paste0("hictools_version=", as.character(packageVersion("hictools"))),
  paste0("bedtorch_version=", as.character(packageVersion("bedtorch"))),
  # All items in script_args
  names(script_args) %>% purrr::map_chr(function(name) {
    v <- script_args[[name]]
    v_str <- if (!is.null(v))
      v %>% purrr::map_chr(as.character) %>% paste(collapse = ":")
    else
      ""
    paste0(name, "=", v_str)
  })
)

library(here)

logging::loginfo(str_interp("Argument summary:"))
comments %>% purrr::walk(function(v) logging::loginfo(v))

all_chroms <- system(str_interp("tabix -l ${script_args$input}"), intern = TRUE)
logging::loginfo(
  str_interp(
    "Found ${length(all_chroms)} chromosomes: ${paste(all_chroms, collapse = \",\")}"
  )
)

# Filter chroms
if (!is.null(script_args$chroms))
  all_chroms <- intersect(all_chroms, script_args$chroms)
if (!is.null(script_args$exclude_chroms))
  all_chroms <- setdiff(all_chroms, script_args$exclude_chroms)
logging::loginfo(
  str_interp(
    "Process ${length(all_chroms)} chromosomes: ${paste(all_chroms, collapse = \",\")}"
  )
)

# Standard compartment
if (is_null(script_args$standard_compartment)) {
  standard_comp <- NULL
} else{
  standard_comp <- local({
    res_kbp <- as.integer(script_args$res / 1e3L)
    comp_name <- script_args$standard_compartment
    if (comp_name == "gene_density")
      comp_name <-
      str_interp("gencode.v30.b37.gene_density.${res_kbp}kbp")
    else if (comp_name == "wbc")
      comp_name <-
      str_interp("wbc.rep1.compartment.hs37-1kg.NONE.${res_kbp}kbp")
    
    logging::loginfo(str_interp("Loading standard compartment: ${comp_name}"))
    env <- rlang::env()
    data(list = comp_name, envir = env, package = "hictools")
    env[[comp_name]]
  })
}

cofrag_comp_grid <- expand_grid(
  chrom = all_chroms,
  method = script_args$method,
  smooth = script_args$smooth,
  # metric = c("score", "hellinger")
  metric = "score"
) 

cofrag_comp_results <- cofrag_comp_grid %>%
  pmap(function(chrom, method, smooth, metric) {
    logging::loginfo(str_interp("Processing: chrom:${chrom} ${method} smooth:${smooth} ${metric}"))
    hic_data <-
      hictools::load_hic_genbed(
        script_args$input,
        chrom,
        resol = script_args$res,
        type = "observed",
        norm = "NONE",
        genome = script_args$genome,
        score_col = metric,
        scale_score = TRUE
      )
    
    comps <-
      hictools::get_compartment(hic_data,
                                method = method,
                                standard = standard_comp,
                                smooth = smooth,
                                genome = script_args$genome)
    
    comps$method <- method
    comps$smooth <- smooth
    comps$metric <- metric
    
    comps$score <- local({
      score <- comps$score
      ifelse(is.infinite(score) | is.na(score), NA, score)
    })
    
    seqinfo <- bedtorch::get_seqinfo(script_args$genome)
    suppressWarnings({
      GenomeInfoDb::seqlevels(comps) <- GenomeInfoDb::seqlevels(seqinfo)
      GenomeInfoDb::seqinfo(comps) <- bedtorch::get_seqinfo(script_args$genome)
      GenomicRanges::trim(comps)
    })
    
    if (!is.null(standard_comp))
      correlation <- hictools::comp_correlation(comps, standard_comp)
    else
      correlation <- NA

    list(comps = comps, correlation = correlation)
  })

comps <- cofrag_comp_results %>% map(~ .$comps) %>% do.call(what = c, args = .)

# cofrag_correlation <-
#   map2_dfr(seq.int(nrow(cofrag_comp_grid)), 
#            cofrag_comp_results, 
#            function(grid_idx, result) {
#     c(cofrag_comp_grid[grid_idx,], result$correlation)
#   })

system(str_interp("mkdir -p ${script_args$output_dir}"))
cofrag_comp_file <- str_interp("${script_args$output_dir}/${script_args$sample_id}.compartment.bed.gz")
cofrag_correlation_file <- str_interp("${script_args$output_dir}/${script_args$sample_id}.compartment_correlation.tsv")
logging::loginfo(str_interp("Write compartment scores: ${cofrag_comp_file}"))
logging::loginfo(str_interp("Write compartment correlation: ${cofrag_correlation_file}"))

bedtorch::write_bed(comps, file_path = cofrag_comp_file, comments = comments, tabix_index = FALSE)
# write_tsv(cofrag_correlation, file = cofrag_correlation_file)

logging::loginfo(str_interp("Analysis completed, with results located at ${script_args$output_dir}"))






inst/extdata/scripts/cofragr.R
#!/usr/bin/env Rscript

stop_quietly <- function() {
  opt <- options(show.error.messages = FALSE)
  on.exit(options(opt))
  stop()
}

# script_args validity
validate_args <- function(args) {
  with(args, {
    assert_that(is_scalar_character(metrics) && metrics %in% c("ks"), msg = "metrics must be ks")
    assert_that(
      is_scalar_character(genome) &&
        genome %in% c("hs37-1kg", "GRCh37", "GRCh38"),
      msg = str_interp("Unsupported genome: ${genome}")
    )
    assert_that(is_scalar_integer(res) && res > 0)
    assert_that(is_scalar_integer(block_size) && block_size > 0 && block_size %% res == 0, msg = "block_size must be a positive integer which is the multiple of res")
    assert_that(is_scalar_integer(ncores) && ncores >= 1)
    assert_that(is_scalar_integer(bootstrap) && bootstrap >= 1)
    assert_that(is_scalar_integer(subsample) && subsample >= 100, msg = "subsample must be a positive integer no less than 100")
    assert_that(is_scalar_integer(min_mapq) && min_mapq >= 0)
    assert_that(is_scalar_integer(min_fraglen) && min_fraglen >= 1)
    assert_that(is_scalar_integer(max_fraglen) && max_fraglen >= 1)
    assert_that(is_null(seed) || is_scalar_integer(seed))
    assert_that(!isTRUE(max_fraglen <= min_fraglen), msg = "max_fraglen must be larger than min_fraglen")
  }) %>% invisible()
} 


# example
script_args <- list(
  input = "~/dev/gateway_project/sandbox/frag/Pilot2_9.hg19.frag.bed.gz",
  output_dir = here::here("sandbox/"),
  sample_id = "Pilot2_9.hg19",
  metrics = "ks",
  genome = "hs37-1kg",
  res = 500e3L,
  block_size = 10e6L,
  ncores = 2L,
  bootstrap = 5L,
  subsample = 10e3L,
  seed = 1228L,
  chroms = c("21"),
  exclude_chroms = NULL,
  min_mapq = 30L,
  min_fraglen = 100L,
  max_fraglen = 350L,
  intersect_region = NULL, # here("sandbox/cofrag/cfEW1.cna.neutral.bed"),
  exclude_region = "encode.blacklist",
  parallel = TRUE
)


if (interactive()) {
  library(tidyverse)
  library(magrittr)
  library(here)
  library(rlang)
  library(assertthat)

  if (is.null(get0("script_args")))
    stop_quietly()

  # Build a string to record all parameters
  param_str <-
    paste(names(script_args),
          script_args,
          sep = "=",
          collapse = "; ")
} else {
  library(here)
  library(rlang)
  library(assertthat)
  
  # Run in CLI script mode
  parser <- optparse::OptionParser(
    option_list = list(
      optparse::make_option(c("-i", "--input")),
      optparse::make_option(c("-o", "--output-dir")),
      optparse::make_option(c("-s", "--sample-id")),
      optparse::make_option(c("-m", "--metrics"), default = "ks"),
      optparse::make_option(c("-g", "--genome"), default = "hs37-1kg",
                            help = "Reference genome for the input"),
      optparse::make_option(c("--res"), type = "integer", default = 500e3L),
      optparse::make_option(c("--block-size"), type = "integer", default = 10e6L),
      optparse::make_option(c("-n", "--ncores"), type = "integer", default = 1L),
      optparse::make_option(c("--bootstrap"), type = "integer", default = 1L),
      optparse::make_option(c("--subsample"), type = "integer", default = NULL),
      optparse::make_option(c("--seed"), type = "integer", default = NULL),
      optparse::make_option(c("--chroms"), default = NULL,
                            help = "Perform the analysis only for a selected group of chromosomes. Separated by colons, such as 12:16:X. If not provided, all chromosomes found in the input file will be used"),
      optparse::make_option(c("--exclude-chroms"), default = NULL,
                            help = "Exclude chromosomes from the analysis. Separated by colons, such as 12:16:X"),
      optparse::make_option(
        c("--min-mapq"),
        type = "integer",
        default = 1L,
        help = "Minimal MAPQ for fragments to take part in the analysis"
      ),
      optparse::make_option(
        c("--min-fraglen"),
        type = "integer",
        default = 100L,
        help = "Minimal length for fragments to take part in the analysis"
      ),
      optparse::make_option(
        c("--max-fraglen"),
        type = "integer",
        default = 350L,
        help = "Maximal length for fragments to take part in the analysis"
      ),
      optparse::make_option(
        c("--intersect-region"),
        type = "character",
        default = NULL,
        help = "BED files defining regions to be intersected with the fragment data file, separated by colon"
      ),
      optparse::make_option(
        c("--exclude-region"),
        type = "character",
        default = NULL,
        # default = "encode.blacklist",
        help = "BED files defining regions to be excluded from the analysis, separated by colon"
      ),
      optparse::make_option(c("--parallel"), default = TRUE, help = "Run in parallel mode. Default is TRUE")
    )
  )
  script_args <-
    optparse::parse_args(
      parser,
      args = commandArgs(trailingOnly = TRUE),
      convert_hyphens_to_underscores = TRUE
    )

  library(tidyverse)
  library(magrittr)

  if (is.null(script_args$seed)) {
    script_args$seed <- as.integer(round(runif(1) * 1e6L))
    logging::loginfo(str_interp("Randomly pick a seed: ${script_args$seed}"))
  }
  if (!is.null(script_args$chroms))
    script_args$chroms %<>% str_split(pattern = ":") %>% .[[1]]
  if (!is.null(script_args$exclude_chroms))
    script_args$exclude_chroms %<>% str_split(pattern = ":") %>% .[[1]]
  if (!is.null(script_args$intersect_region))
    script_args$intersect_region %<>% str_split(pattern = ":") %>% .[[1]]
  if (!is.null(script_args$exclude_region))
    script_args$exclude_region %<>% str_split(pattern = ":") %>% .[[1]]

  # Build a string to record all parameters
  param_str <- paste0("Parameters: ", paste(commandArgs(trailingOnly = TRUE), collapse = " "))
}

validate_args(script_args)

# Build comment lines
comments <- c(
  paste0("cofragr_version=", as.character(packageVersion("cofragr"))),
  paste0("hictools_version=", as.character(packageVersion("hictools"))),
  paste0("bedtorch_version=", as.character(packageVersion("bedtorch"))),
  # All items in script_args
  names(script_args) %>% purrr::map_chr(function(name) {
    v <- script_args[[name]]
    v_str <- if (!is.null(v))
      v %>% purrr::map_chr(as.character) %>% paste(collapse = ":")
    else
      ""
    paste0(name, "=", v_str)
  })
)

logging::loginfo(str_interp("Argument summary:"))
comments %>% purrr::walk(function(v) logging::loginfo(v))

all_chroms <- system(str_interp("tabix -l ${script_args$input}"), intern = TRUE)
logging::loginfo(
  str_interp(
    "Found ${length(all_chroms)} chromosomes: ${paste(all_chroms, collapse = \",\")}"
  )
)
# Filter chroms
if (!is_null(script_args$chroms))
  all_chroms <- intersect(all_chroms, script_args$chroms)
if (!is_null(script_args$exclude_chroms))
  all_chroms <- setdiff(all_chroms, script_args$exclude_chroms)
logging::loginfo(
  str_interp(
    "Process ${length(all_chroms)} chromosomes: ${paste(all_chroms, collapse = \",\")}"
  )
)


# Users may provide multiple files for excluded regions, and for each file the regions may be overlapping.
# Merge all excluded regions together.
merge_regions <- function(regions, genome = NULL) {
  if (is_null(regions))
    return(NULL)

  regions %>% map(function(r) {
    # Use regions shipped with the package
    if (r == "encode.blacklist")
      local({
        env <- rlang::env()
        dataset_name <- paste0("wgEncodeDacMapabilityConsensusExcludable.", genome)
        data(list = dataset_name, package = "cofragr", envir = env)
        region <- env[[dataset_name]]
        assert_that(!is_null(region))
        region
      })
    else
      bedtorch::read_bed(r, genome = genome)
  }) %>%
    do.call(what = c, args = .) %>%
    bedtorch::merge_bed()
}


filter_mapq <- function(frag, min_mapq) {
  if (length(frag) == 0) {
    logging::loginfo("filter_mapq: empty fragments, skipping")
    return(frag)
  }

  if (!is.null(min_mapq)) {
    frag <- frag[frag$mapq >= min_mapq]
    # logging::loginfo(str_interp("# of fragments after MAPQ filter: ${length(frag)}"))
  }
  frag
}

filter_fraglen <- function(frag, min_fraglen = NULL, max_fraglen = NULL) {
  if (length(frag) == 0) {
    logging::loginfo("filter_length: empty fragments, skipping")
    return(frag)
  }

  if (is.null(min_fraglen) && is.null(max_fraglen))
    return(frag)
  else if (is.null(min_fraglen))
    frag <- frag[frag$len <= max_fraglen]
  else if (is.null(max_fraglen))
    frag <- frag[frag$len >= min_fraglen]
  else
    frag <- frag[between(frag$len, min_fraglen, max_fraglen)]

  frag
}

cofrag_cm <- all_chroms %>% map(function(chrom) {
  logging::loginfo(str_interp("Loading fragments for chromosome ${chrom}"))

  frag <- cofragr::read_fragments(file_path = script_args$input, range = chrom, genome = script_args$genome)

  # if ("cigar1" %in% frags_cols && "cigar2" %in% frags_cols) {
  #   # Exclude soft clipping
  #   frags %<>%
  #     filter(!(str_detect(cigar1, pattern = "^[0-9]+S") | str_detect(cigar2, pattern = "[0-9]+S$")))
  # }

  logging::loginfo(str_interp("# of fragments loaded for chr${chrom}: ${length(frag)}"))

  if (!is_null(script_args$intersect_region)) {
    intersect_region <- merge_regions(script_args$intersect_region, genome = script_args$genome)
    frag %<>% bedtorch::intersect_bed(intersect_region, mode = "unique")
    logging::loginfo(str_interp("# of fragments after intersecting regions: ${length(frag)}"))
  }

  if (!is_null(script_args$exclude_region)) {
    exclude_region <- merge_regions(script_args$exclude_region, genome = script_args$genome)
    frag %<>% bedtorch::exclude_bed(exclude_region)
    logging::loginfo(str_interp("# of fragments after excluding regions: ${length(frag)}"))
  }

  if (!is_null(script_args$min_mapq)) {
    frag %<>% filter_mapq(min_mapq = script_args$min_mapq)
    logging::loginfo(str_interp("# of fragments after filtering by MAPQ: ${length(frag)}"))
  }

  if (!is_null(script_args$min_fraglen) || !is_null(script_args$max_fraglen)) {
    frag %<>% filter_fraglen(min_fraglen = script_args$min_fraglen, max_fraglen = script_args$max_fraglen)
    logging::loginfo(str_interp("# of fragments after filtering by fragment length: ${length(frag)}"))
  }

  # Minimal # of fragments: 10,000
  if (length(frag) < 10e3L) {
    logging::logwarn(str_interp("Not enough fragments to call the matrix: ${length(frag)}"))
    return(NULL)
  }

  fraglen_list <- cofragr::preprocess_frag_bed(frag, bin_size = script_args$res)
  # Explicitly remove frag to save memory
  rm(frag)
  gc()

  logging::loginfo(str_interp("Completed preprocessing. Current memory usage: ${as.numeric(lobstr::mem_used()) / 1e6}"))

  cofragr::contact_matrix(
    fraglen_list,
    bin_size = script_args$res,
    block_size = script_args$block_size,
    n_workers = script_args$ncores,
    parallel = script_args$parallel,
    subsample = script_args$subsample,
    min_sample_size = 100L,
    bootstrap = script_args$bootstrap,
    seed = script_args$seed
  )
}) %>%
  hictools::concat_hic()
  # do.call(what = c, args = .)


output_dir <- script_args$output_dir
system(str_interp("mkdir -p ${output_dir}"))
cofrag_cm_file <- str_interp("${output_dir}/${script_args$sample_id}.cofrag_cm.bed.gz")
logging::loginfo(str_interp("Write cofragmentation matrix: ${cofrag_cm_file}"))

hictools::write_hic_bedtorch(cofrag_cm, file_path = cofrag_cm_file, comments = comments)
# cofragr::write_contact_matrix(cofrag_cm, file_path = cofrag_cm_file, comments = comments)

logging::loginfo(str_interp("Analysis completed, with results located at ${output_dir}"))


